File: _examples/C/full_integration_step.c
/*
 * Copyright Core DF
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Core Auto real-time step — full integration example (C port).
 * Build: make
 */
#include "../../cawbs/C/include/cawbs.h"
#include "../../files/C/include/fileclient.h"
#include <cJSON.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void die(const char *msg) {
    fprintf(stderr, "%s\n", msg);
    exit(1);
}

static int json_status(const char *json) {
    cJSON *r = cJSON_Parse(json);
    if (!r) return 0;
    cJSON *sc = cJSON_GetObjectItem(r, "status_code");
    int code = sc && cJSON_IsNumber(sc) ? (int)sc->valuedouble : 0;
    cJSON_Delete(r);
    return code;
}

int main(void) {
    wbs_result r = cawbs_init();
    if (r.status_code != 200) die("cawbs_init failed");
    wbs_result_free(&r);

    r = cawbs_get_event_payload();
    if (r.status_code != 200) die("cawbs_get_event_payload failed");
    wbs_result_free(&r);

    const char *ack_dir = getenv("EXAMPLE_ACK_DIR");
    if (!ack_dir) ack_dir = "/tmp/coreauto-example";
    char path[512];
    snprintf(path, sizeof(path), "%s/unknown.json", ack_dir);

    char *wr = file_local_write(path, "{\"orderId\":\"unknown\"}");
    if (!wr || json_status(wr) != 200) die("file_local_write failed");
    free(wr);

    char out[640];
    snprintf(out, sizeof(out), "{\"orderId\":\"unknown\",\"ackPath\":\"%s\"}", path);
    r = cawbs_put_step_payload(out);
    if (r.status_code != 200) die("cawbs_put_step_payload failed");
    wbs_result_free(&r);

    printf("{\"status_code\":200,\"result\":%s}\n", out);
    return 0;
}

File: _examples/Rust/src/full_integration_step.rs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Core Auto real-time step — full integration example (Rust port).

use cawbs::{Cawbs, Result as WbsResult};
use coreauto_files::local_write;
use coreauto_kafka::produce;
use coreauto_transform::json_stringify;
use serde_json::{json, Value};

fn must(label: &str, r: WbsResult) {
    if r.status_code >= 400 || r.status_code == 0 {
        eprintln!("{label}: {r:?}");
        std::process::exit(1);
    }
}

fn status_ok(v: &Value) -> bool {
    v.get("status_code").and_then(|c| c.as_i64()) == Some(200)
}

fn optional(label: &str, v: Value) -> bool {
    let code = v.get("status_code").and_then(|c| c.as_i64()).unwrap_or(0);
    let err = v.get("error").and_then(|e| e.as_str()).unwrap_or("").to_lowercase();
    if [601, 500].contains(&code) && err.contains("missing") {
        println!("[skip] {label}: not configured");
        return false;
    }
    if code >= 400 || code == 0 {
        println!("[warn] {label}: {err}");
        return false;
    }
    println!("[ok] {label}");
    true
}

fn order_id(event: &WbsResult) -> String {
    event
        .payload
        .as_ref()
        .and_then(|p| p.get("orderId").or_else(|| p.get("id")))
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string()
}

fn main() {
    must("cawbs.init", Cawbs::init());
    let event = Cawbs::get_event_payload();
    must("cawbs.get_event_payload", event.clone());

    let order_id = order_id(&event);
    let ack_dir =
        std::env::var("EXAMPLE_ACK_DIR").unwrap_or_else(|_| "/tmp/coreauto-example".into());
    let ack_path = format!("{ack_dir}/{order_id}.json");

    let order = json!({ "orderId": order_id, "details": event.payload });
    let text = json_stringify(&order, None);
    if !status_ok(&text) {
        eprintln!("transform failed: {text}");
        std::process::exit(1);
    }
    let body = text.get("text").and_then(|v| v.as_str()).unwrap_or("");
    let wr = local_write(&ack_path, body);
    if !status_ok(&wr) {
        eprintln!("write failed: {wr}");
        std::process::exit(1);
    }

    let topic = std::env::var("EXAMPLE_KAFKA_TOPIC").unwrap_or_else(|_| "orders.enriched".into());
    let mut published = Vec::new();
    if optional(
        "queues.kafka",
        produce(&topic, order.clone(), None),
    ) {
        published.push("kafka");
    }

    let out = json!({
        "orderId": order_id,
        "queuesPublished": published,
        "ackPath": ack_path
    });
    must("cawbs.put_step_payload", Cawbs::put_step_payload(out.clone()));
    println!("{}", json!({ "status_code": 200, "result": out }));
}
